clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -1259,16 +1263,48 @@ raw2trace_t::append_bb_entries(raw2trace_thread_data_t *tdata,
         DR_CHECK(pc > decode_pc, "error advancing inside block");
         DR_CHECK(!instr->is_cti() || i == instr_count - 1, "invalid cti");
+        if (instr->is_syscall() && should_omit_syscall(tdata)) {
+            add_to_statistic(tdata, RAW2TRACE_STAT_FALSE_SYSCALL, 1);
+            log(3, "Omitting syscall instr without subsequent number marker.\n");
+            // Exit and do not append this syscall instruction.  It must be the
+            // final instruction in the block; since the tracer requests callbacks
+            // on all syscalls, none are inlined.
+            DR_CHECK(i == instr_count - 1, "syscall not last in block");
+            break;
+        }
         // TODO i#5934: This is a workaround for the trace invariant error triggered
         // by consecutive duplicate system call instructions. Remove this when the
         // error is fixed in the drmemtrace tracer.
+        // TODO i#6102: This actually does the wrong thing for SIG_IGN interrupting
+        // an auto-restart syscall; we live with that until we remove it after fixing
+        // the incorrect duplicate syscall error.
         if (instr->is_syscall() && get_last_pc_if_syscall(tdata) == orig_pc &&
             instr_count == 1) {
+            // Also remove the syscall marker.  It could be after a timestmap;cpuid
+            // pair; we're fine removing those too and having the prior timestamp
+            // go with the next block.
+            if (TESTANY(OFFLINE_FILE_TYPE_SYSCALL_NUMBERS, tdata->file_type)) {
+                const offline_entry_t *entry = get_next_entry(tdata);
+                if (entry->timestamp.type == OFFLINE_TYPE_TIMESTAMP) {
+                    entry = get_next_entry(tdata);
+                    if (entry->extended.type == OFFLINE_TYPE_EXTENDED &&
+                        entry->extended.ext == OFFLINE_EXT_TYPE_MARKER &&
+                        entry->extended.valueB == TRACE_MARKER_TYPE_CPU_ID) {
+                        entry = get_next_entry(tdata);
+                    }
+                }
+                DR_CHECK(entry->extended.type == OFFLINE_TYPE_EXTENDED &&
+                             entry->extended.ext == OFFLINE_EXT_TYPE_MARKER &&
+                             entry->extended.valueB == TRACE_MARKER_TYPE_SYSCALL,
+                         "Syscall without marker should have been removed");
+                // We've consumed these records and we just drop them.
+            }
             add_to_statistic(tdata, RAW2TRACE_STAT_DUPLICATE_SYSCALL, 1);
-            log(3, "Found block with duplicate system call instruction. Skipping.");
+            log(3, "Found block with duplicate system call instruction. Skipping.\n");
             // Since this instr is in its own block, we're done.
             // Note that this will result in a pair of timestamp+cpu markers without
-            // anything before the next timestamp+cpu pair.
+            // anything before the next timestamp+cpu pair for legacy traces w/o
```

Comment:
I don't understand "syscall-in-middle-of-block-delivered-via-dispatch case": a syscall with drmemtrace will never be in the middle of a block due to the filter event always returning true.

Reply:
Sorry, that's a typo. I meant signal-in-the-middle-of-block (the first scenario in should_omit_syscall)

Reply:
Still confused:

"The timstamp+cpu will be present also": you mean if there is a duplicate syscall removal?  But the should_omit_syscall-#1 will have the syscall removed so there won't be any dup syscall.  Not seeing how this relates to line 1306 talking about dup syscall removal?

Reply:
L1306 says that the timestamp+cpu marker without anything in that block will happen only for legacy traces. But I think it'll happen in current traces too. For the scenario where there's a signal in the middle of the block (which will give us the syscall instr without the syscall marker).

From the other comment: "Maybe add another test like the previous test_duplicate_syscalls one but without the syscall_num marker after the first syscall?"

Reply:
So something like:
```
   raw.push_back(make_header());
    raw.push_back(make_tid());
    raw.push_back(make_pid());
    raw.push_back(make_line_size());
    raw.push_back(make_timestamp(1));
    raw.push_back(make_core());
    raw.push_back(make_block(offs_move1, 2));
    // remove this one -- raw.push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL, SYSCALL_NUM));
    raw.push_back(make_timestamp(2));
    raw.push_back(make_core());
    // Repeat the syscall that was the second instr in the size-2 block above,
    // in its own separate block. This is the signature of the duplicate
    // system call invariant error seen in i#5934.
    raw.push_back(make_block(offs_sys, 1));
    // New traces have a syscall marker, of which we test removal.
    raw.push_back(make_timestamp(3));
    raw.push_back(make_core());
    raw.push_back(make_marker(TRACE_MARKER_TYPE_SYSCALL, SYSCALL_NUM));
    raw.push_back(make_timestamp(4));
    raw.push_back(make_core());
    raw.push_back(make_block(offs_move2, 1));
    raw.push_back(make_exit());
```

Reply:
> But the should_omit_syscall-#1 will have the syscall removed so there won't be any dup syscall. 

Oh wait. You're saying it will remove the first syscall as a false syscall. So it won't cause any issues? Maybe. Better to add a test still.

Reply:
My concern was coming from clients/drcachesim/tracer/raw2trace.cpp:L1286 which removes the timestamp+cpu only if there is a syscall num marker following the syscall instr. Could it happen that there's no marker? If yes, then we won't remove the following timestamp+cpu?

Reply:
A syscall with no following marker is removed as a "false" syscall before we look for any duplicate syscall to remove so no it should not happen.  The false syscall unit test has a syscall with no marker following by a syscall (with timestamp in between) so I believe it already looks like your suggested test above (after I removed the pre-syscall instr so now it's a bare syscall).

Reply:
>  The false syscall unit test has a syscall with no marker following by a syscall (with timestamp in between)

But the false syscall test repeated the move1 too between the two syscalls:

```
raw.push_back(make_block(offs_move1, 2));
    // There is no syscall marker here, so the syscall should be removed.
    raw.push_back(make_timestamp(2));
    raw.push_back(make_core());
    // Repeat the syscall but with a marker this time.
    // Start with the move to not trigger dup syscall removal.  <--- this one
    raw.push_back(make_block(offs_move1, 2));
```

> after I removed the pre-syscall instr so now it's a bare syscall

Oh you mean you changed it but it's not yet pushed?

Reply:
Right hadn't pushed any review changes yet.  Pushed now.